rest_framework/schemas/openapi.py

```diff
@@ -479,29 +493,67 @@ def _get_serializer(self, method, path):
                           .format(view.__class__.__name__, method, path))
             return None
 
-    def _get_request_body(self, path, method):
-        if method not in ('PUT', 'PATCH', 'POST'):
+    class SchemaMode(Enum):
+        RESPONSE = 1
+        BODY = 2
+
+    def _get_item_schema(self, serializer, schema_mode, method):
+        if not isinstance(serializer, serializers.Serializer):
             return {}
 
-        self.request_media_types = self.map_parsers(path, method)
+        # If the serializer uses a model, we should use a reference
+        if hasattr(serializer, 'Meta') and hasattr(serializer.Meta, 'model'):
+            model_name = serializer.Meta.model.__name__
+            return {'$ref': '#/components/schemas/{}'.format(model_name)}
+
+        # There is no model, we'll map the serializer's fields
+        item_schema = self._map_serializer(serializer)
 
+        if schema_mode == self.SchemaMode.RESPONSE:
+            # No write_only fields for response.
+            for name, schema in item_schema['properties'].copy().items():
+                if 'writeOnly' in schema:
+                    del item_schema['properties'][name]
+                    if 'required' in item_schema:
+                        item_schema['required'] = [f for f in item_schema['required'] if f != name]
+
+        elif schema_mode == self.SchemaMode.BODY:
+            # No required fields for PATCH
+            if method == 'PATCH':
+                item_schema.pop('required', None)
+            # No read_only fields for request.
+            for name, schema in item_schema['properties'].copy().items():
+                if 'readOnly' in schema:
+                    del item_schema['properties'][name]
+
+        return item_schema
+
+    def _get_component_schema(self, path, method):
         serializer = self._get_serializer(path, method)
 
         if not isinstance(serializer, serializers.Serializer):
-            return {}
+            return None
+
+        # If the model has no model, then the serializer will be inlined
+        if not hasattr(serializer, 'Meta') or not hasattr(serializer.Meta, 'model'):
+            return None
```

Comment:
Q: Do we need to allow for different components (with different fields) for requests and responses, at least?

Reply:
Good question ! We could have have two components like `ModelNameRequest` and `ModelNameResponse`.
However, we may better keep one component and use the `readOnly` and `writeOnly` properties.

This is the generator's job to deduce that : 
- properties having `writeOnly: true` should not be in a Response
- properties having `readOnly: true` should not be in Request

More information here : https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md